### PR TITLE
Use images as charts

### PR DESCRIPTION
**Pull request by robinschmid**
- create class to buffer
- first click is now recognized on BorderPane
- resize only works when making the column bigger... but at least something

### CODE REVIEW

File: src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/BufferedChartNode.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  private static final Logger logger = Logger.getLogger(BufferedChartNode.class.getName());
+  private ChartViewer chart;
+  private ImageView imageView;
+  private boolean makeInteractiveOnClick;
+
+  public BufferedChartNode(final boolean makeInteractiveOnClick) {
+    this.makeInteractiveOnClick = makeInteractiveOnClick;
+    PauseTransition resizer = new PauseTransition(Duration.seconds(2));
+
+    widthProperty().addListener((observable, oldValue, newValue) -> {
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/BufferedChartNode.java:52]: I think the nice thing about the image is, is that we can do all the calculations on a seperate thread, including creation of the chart and the image. When the column is changed now, all images are recalcuated on the FX thread, or?

File: src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeMobilogramChart.java [hunk 2]
```diff
@@ -107,16 +103,7 @@ public FeatureShapeMobilogramChart(@NotNull ModularFeatureListRow row, AtomicDou
...
-    ImageView view = new ImageView(SwingFXUtils.toFXImage(img, null));
-    view.setOnMouseClicked(e -> MZmineCore.runLater(() -> {
-      getChildren().remove(view);
-      getChildren().add(chart);
-      e.consume();
-    }));
-
-    MZmineCore.runLater(() -> getChildren().add(view));
+    setChartCreateImage(chart, GraphicalColumType.DEFAULT_GRAPHICAL_CELL_WIDTH,
+        GraphicalColumType.DEFAULT_IMAGE_CELL_HEIGHT);
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeMobilogramChart.java]: GraphicalColumType.DEFAULT_GRAPHICAL_CELL_HEIGHT, mobilograms dont need to be as high

File: src/main/java/io/github/mzmine/datamodel/features/types/FeatureShapeType.java [hunk 3]
```diff
@@ -91,7 +90,7 @@ public Node getCellNode(TreeTableCell<ModularFeatureListRow, Boolean> cell,
 
   @Override
   public double getColumnWidth() {
-    return DEFAULT_GRAPHICAL_CELL_WIDTH;
+    return LARGE_GRAPHICAL_CELL_WIDTH;
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/datamodel/features/types/FeatureShapeType.java:93]: why change this now?

File: src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeChart.java [hunk 4]
```diff
@@ -105,23 +100,14 @@ public FeatureShapeChart(@NotNull ModularFeatureListRow row, AtomicDouble progre
...
-            GraphicalColumType.DEFAULT_GRAPHICAL_CELL_HEIGHT);
-    ImageView view = new ImageView(SwingFXUtils.toFXImage(img, null));
-    view.setOnMouseClicked(e -> MZmineCore.runLater(() -> {
-      getChildren().remove(view);
-      getChildren().add(chart);
-      e.consume();
-    }));
-
-    MZmineCore.runLater(() -> getChildren().add(view));
+    var width = GraphicalColumType.LARGE_GRAPHICAL_CELL_WIDTH;
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeChart.java:108]: why change?

**Review by SteffenHeu — COMMENTED** [review]
Some questions

File: src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeIonMobilityRetentionTimeHeatMapChart.java [hunk 5]
```diff
@@ -81,23 +77,8 @@ public FeatureShapeIonMobilityRetentionTimeHeatMapChart(@NotNull ModularFeature
...
-    }));
-
-    Platform.runLater(() -> getChildren().add(view));
-
-//    chart.addDatasetsChangedListener(
-//        (e) -> Platform.runLater(() -> chart.getXYPlot().getRangeAxis().setAutoRange(true)));
-//    getChildren().add(chart);
-//    Platform.runLater(() -> chart.setDataset(dataset));
+    setChartCreateImage(chart, GraphicalColumType.LARGE_GRAPHICAL_CELL_WIDTH,
+        GraphicalColumType.DEFAULT_IMAGE_CELL_HEIGHT);
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeIonMobilityRetentionTimeHeatMapChart.java]: i dont think this needs to be image height

File: src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/BufferedChartNode.java (hunk 1 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/BufferedChartNode.java:52]: They are only recalculated once you scroll down and see the images. We could abstract this part from the jfx thread and then push the change on the JavaFX thread I guess

File: src/main/java/io/github/mzmine/datamodel/features/types/FeatureShapeType.java (hunk 3 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/datamodel/features/types/FeatureShapeType.java:93]: The chromatograms can be quite wide and its better to start out with a wider column. Better to see if there is a shoulder on a peak or if there are multiple peaks in a chromatogram

File: src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeMobilogramChart.java (hunk 2 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeMobilogramChart.java]: True copy paste error

Comment (robinschmid) [reply src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeMobilogramChart.java]: changed back to DEFAULT_GRAPHICAL_CELL_HEIGHT

File: src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeChart.java (hunk 4 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeChart.java:108]: The chromatograms can be quite wide and its better to start out with a wider column. Better to see if there is a shoulder on a peak or if there are multiple peaks in a chromatogram

File: src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeIonMobilityRetentionTimeHeatMapChart.java (hunk 5 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeIonMobilityRetentionTimeHeatMapChart.java]: true not sure why I missed this

Comment (robinschmid) [reply src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeIonMobilityRetentionTimeHeatMapChart.java]: reverted to 
DEFAULT_GRAPHICAL_CELL_HEIGHT

File: src/main/java/io/github/mzmine/datamodel/features/types/FeatureShapeType.java (hunk 3 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/features/types/FeatureShapeType.java:93]: Making the chromatogram plot wider won't change that, as the visible X axes range is determined by the fwhm of the feature.

File: src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeChart.java (hunk 4 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/features/types/graphicalnodes/FeatureShapeChart.java:108]: See above reply

File: src/main/java/io/github/mzmine/datamodel/features/types/FeatureShapeType.java (hunk 3 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/datamodel/features/types/FeatureShapeType.java:93]: Making it wider will make the peak displayed wider. That's what I mean

**Comment by SteffenHeu** [comment]
Looks good, thanks

**Pull request merged**